### PR TITLE
Migrating HTML reprs to jinja2

### PR DESCRIPTION
**Pull request by jacobtomlinson**
This PR starts the task of migrating HTML reprs over to use Jinja2 templates instead of multiline strings. There should be no changes to the reprs themselves but they should be easier to create and more maintainable going forwards.

**Before**

Currently we construct HTML within Python strings

```python

class HasWhat(dict):
    """A dictionary of all workers and which keys that worker has."""

    def _repr_html_(self):
        rows = ""

        for worker, keys in sorted(self.items()):
            summary = ""
            for key in keys:
                summary += f"""<tr><td>{key}</td></tr>"""

            rows += f"""<tr>
            <td>{worker}</td>
            <td>{len(keys)}</td>
            <td>
                <details>
                <summary style='display:list-item'>Expand</summary>
                <table>
                {summary}
                </table>
                </details>
            </td>
        </tr>"""

        output = f"""
        <table>
        <tr>
            <th>Worker</th>
            <th>Key count</th>
            <th>Key list</th>
        </tr>
        {rows}
        </table>
        """

        return output
```

**After**

Instead we can use the new `distributed.widgets` submodule to render this from a template.

```python
from distributed.widgets import get_template

class HasWhat(dict):
    """A dictionary of all workers and which keys that worker has."""

    def _repr_html_(self):
        return get_template("has_what.html.j2").render(has_what=self)
```

And the template itself gets placed in `distributed/widgets/templates` and is much more readable.

```jinja2
<table>
<tr>
    <th>Worker</th>
    <th>Key count</th>
    <th>Key list</th>
</tr>
{% for worker, keys in has_what.items() %}
    <tr>
        <td>{{ worker }}</td>
        <td>{{ keys | length }}</td>
        <td>
            <details>
            <summary style='display:list-item'>Expand</summary>
            <table>
            {% for key in keys %}
                <tr><td>{{ key }}</td></tr>
            {% endfor %}
            </table>
            </details>
        </td>
    </tr>
{% endfor %}
</table>
```

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -1,6 +1,7 @@
 click >= 6.6
 cloudpickle >= 1.5.0
 dask == 2021.07.2
+jinja2
```
Comment (fjetter) [inline requirements.txt:4]: You should add this to the `continuous_integration/environment_py3*.py` files as well. Jinja should already be included as a transitive dependency but the conda solver may create different results if it is a first level dependency.

**Review by fjetter — COMMENTED** [review]
FYI jinja is already a transitive dependency via bokeh. Bokeh, however, is currently an optional dependency so this would promote jinja to a must have. I'm not too worried about this, just wanted to highlight this.

Changes otherwise look good to me

File: requirements.txt (hunk 1 above)
Comment (jacobtomlinson) [reply requirements.txt:4]: Thanks @fjetter. We discussed adding `jinja2` as a top level dependency at the maintainers standup a few months ago and the consensus was to go ahead. I missed the CI environment though.

Comment (jacobtomlinson) [reply requirements.txt:4]: We also have HTML reprs in `dask` which would benefit from this, so we may need to make `jinja2` a dependency of `dask` too.

Comment (fjetter) [reply requirements.txt:4]: I probably wasn't, yet, around during that developer meeting but I don't have any objections to having it as a dependency. I think jinja2 has been around for long enough to consider it stable and it's not very invasive. all good.

The templates are much nicer to read! Also, we could even enable linters and/or formatters for this now :) one step at a time, though.

**Review by fjetter — APPROVED** [review]

File: requirements.txt (hunk 1 above)
Comment (jacobtomlinson) [reply requirements.txt:4]: > Also, we could even enable linters and/or formatters for this now :) one step at a time, though.

Yeah, just having my editor do hightlighting properly is awesome. GitHub too.

We could totally look at introducing prettier though.

File: distributed/client.py [hunk 2]
```diff
@@ -359,6 +358,13 @@ def traceback(self, timeout=None, **kwargs):
     def type(self):
         return self._state.type
 
+    @property
+    def _typ(self):
```
Comment (jrbourbeau) [inline distributed/client.py:362]: Nitpick: Should we call this `_type` instead? Adding the `e` at the end might make this slightly more clear for readers

File: distributed/widgets/__init__.py [hunk 3]
```diff
@@ -0,0 +1 @@
+from .utils import get_environment, get_template  # noqa
```
Comment (jrbourbeau) [inline distributed/widgets/__init__.py:1]: I don't think this is needed as `flake8` is configured to ignore `__init__.py` files

```suggestion
from .utils import get_environment, get_template
```

File: distributed/widgets/templates/client.html.j2 [hunk 4]
```diff
@@ -0,0 +1,47 @@
+<div>
```
Comment (jrbourbeau) [inline distributed/widgets/templates/client.html.j2:1]: I think we need to include an entry to our [`MANIFEST.in` file](https://github.com/dask/distributed/blob/ac55f25d230c7144ed618d1d4374f254303a4e0a/MANIFEST.in) for these templates so they are packaged with `distributed`

File: distributed/client.py (hunk 2 above)
Comment (jacobtomlinson) [reply distributed/client.py]: Done thanks!

File: distributed/widgets/__init__.py (hunk 3 above)
Comment (jacobtomlinson) [reply distributed/widgets/__init__.py]: Done!

File: distributed/widgets/templates/client.html.j2 (hunk 4 above)
Comment (jacobtomlinson) [reply distributed/widgets/templates/client.html.j2:1]: Good point, added all `.j2` files to the manifest.

**Comment by jacobtomlinson** [comment]
Thanks for the review @jrbourbeau. This is now complete, all HTML reprs have been migrated. Ready for final review and merge.

File: distributed/client.py [hunk 5]
```diff
@@ -359,6 +358,13 @@ def traceback(self, timeout=None, **kwargs):
     def type(self):
         return self._state.type
 
+    @property
+    def _type(self):
+        try:
+            return self.type.__module__.split(".")[0] + "." + self.type.__name__
+        except AttributeError:
+            return str(self.type)
```
Comment (mrocklin) [inline distributed/client.py:366]: There might be a `typename` function somewhere

Comment (jacobtomlinson) [reply distributed/client.py:366]: Oooh good point, that's exactly what we want here.

**Review by jrbourbeau — APPROVED** [review]
Thanks @jacobtomlinson! This is in

**Pull request merged**